src/type/validate.js

```diff
@@ -578,6 +581,64 @@ function validateInputFields(
   });
 }
 
+function validateInputObjectCircularReferences(
```

Comment:
Shouldn't this be in a seperate rules file, similar to how NoFragmentCycles is implemented?

Reply:
@spawnia `src/validate/rules/*` is intended to validate AST only with a focus on the query syntax. You can create the schema using SDL or JS classes like here:
```js
new GraphQLSchema({
  queryType: new GraphQLObjectType({
    name: 'Query',
    fields: {
      foo: {
        type: GraphQLString,
      },
    }
  }),
})
```
Also you construct schema from introspection using `buildClientSchema`.
So schema validation shouldn't be tied to AST and should only use `GraphQL*` classes.

On the other hand, your queries are always AST so they validation is also AST based. 

That's why there is two completely different validation mechanisms: `validate`(AST based, mostly query) and `validateSchema`(`GraphQL*` classes, only for schema).